server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -0,0 +1,355 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE PolyKinds #-}
+
+module Hasura.Eventing.ScheduledTrigger
+  ( processScheduledQueue
+  , runScheduledEventsGenerator
+  ) where
+
+import           Control.Concurrent              (threadDelay)
+import           Control.Exception               (try)
+import           Data.Has
+import           Data.IORef                      (IORef, readIORef)
+import           Data.Time.Clock
+import           Hasura.Eventing.HTTP
+import           Hasura.Prelude
+import           Hasura.RQL.DDL.Headers
+import           Hasura.RQL.Types.ScheduledTrigger
+import           Hasura.RQL.Types
+import           Hasura.SQL.DML
+import           Hasura.SQL.Types
+import           System.Cron
+import           Hasura.HTTP
+
+import qualified Data.Aeson                      as J
+import qualified Data.Aeson.Casing               as J
+import qualified Data.Aeson.TH                   as J
+import qualified Data.TByteString                as TBS
+import qualified Data.Text                       as T
+import qualified Database.PG.Query               as Q
+import qualified Hasura.Logging                  as L
+import qualified Network.HTTP.Client             as HTTP
+import qualified Network.HTTP.Types              as HTTP
+import qualified Text.Builder                    as TB (run)
+import qualified Data.HashMap.Strict             as Map
+
+import           Debug.Trace
+
+invocationVersion :: Version
+invocationVersion = "1"
+
+oneSecond :: Int
+oneSecond = 1000000
+
+oneMinute :: Int
+oneMinute = 60 * oneSecond
+
+oneHour :: Int
+oneHour = 60 * oneMinute
+
+type ScheduledEventPayload = J.Value
+
+scheduledEventsTable :: QualifiedTable
+scheduledEventsTable =
+  QualifiedObject
+    hdbCatalogSchema
+    (TableName $ T.pack "hdb_scheduled_events")
+
+data ScheduledEventSeed
+ = ScheduledEventSeed
+ { sesName          :: !TriggerName
+ , sesScheduledTime :: !UTCTime
+ } deriving (Show, Eq)
+
+-- type PartialScheduledEvent = J.Value -> UTCTime -> ScheduledEvent
+
+data SE_P = SE_PARTIAL | SE_FULL
+
+type family Param (p :: k) x
+
+data ScheduledEvent (p :: SE_P)
+  = ScheduledEvent
+  { seId            :: !Text
+  , seName          :: !TriggerName
+  , seScheduledTime :: !UTCTime
+  , seWebhook       :: !(Param p T.Text)
+  , sePayload       :: !(Param p J.Value)
+  }
+
+deriving instance Show (ScheduledEvent 'SE_PARTIAL)
+deriving instance Show (ScheduledEvent 'SE_FULL)
+
+type instance Param 'SE_PARTIAL a = ()
+type instance Param 'SE_FULL a = a
+
+instance J.ToJSON (Param 'SE_PARTIAL a) where
```

Comment:
@lexi-lambda This doesn't compile. Need your help/suggestions here.

Reply:
Trying something fancy here (imitating @jberryman 's idea from Remote Joins)

Reply:
You shouldn’t need this instance or the one for `SE_FULL`. Type families are like type synonyms, so when you write

```haskell
type instance Param 'SE_PARTIAL a = ()
```

that means `Param 'SE_PARTIAL a` *is* `()`. They’re literally the same type from the perspective of GHC. So this instance is equivalent to writing `instance J.ToJSON () where`, which is of course unnecessary, since `()` already has a `ToJSON` instance.

Reply:
> Type families are like type synonyms

(yeah, type families made sense for me once I thought of them as type-level functions (evaluated at compile time), and as a more powerful version of type synonyms (which are natural transformations, maybe))